pyface/data_view/value_types/enum_value.py

```diff
@@ -0,0 +1,162 @@
+#
+# Thanks for using Enthought open source!
+
+from traits.api import Callable, List
+
+from .editable_value import EditableValue
+
+
+class EnumValue(EditableValue):
+    """ Editable value that takes one of a collection of pre-set values.
+
+    Each value can be associated with text, colors and images by supplying
+    functions ``format``, ``colors`` and ``images``, respectively.
+    """
+
+    #: The list of values which are allowed for the value.
+    values = List()
+
+    #: A function that converts a value to a string for display.
+    format = Callable(str, update_value_type=True)
+
+    #: A map from valid values to colors.
+    colors = Callable(None, update_value_type=True)
+
+    #: A map from valid values to images.
+    images = Callable(None, update_value_type=True)
+
+    def is_valid(self, model, row, column, value):
+        """ Whether or not the value is valid for the data item specified.
+
+        The default implementation returns True for all values.
+
+        Parameters
+        ----------
+        model : AbstractDataModel
+            The data model holding the data.
+        row : sequence of int
+            The row in the data model being queried.
+        column : sequence of int
+            The column in the data model being queried.
+        value : any
+            The value to validate.
+
+        Returns
+        -------
+        is_valid : bool
+            Whether or not the value is valid.
+        """
+        return value in self.values
+
+    def has_text(self, model, row, column):
+        """ Get the display text from the underlying value.
+
+        Parameters
+        ----------
+        model : AbstractDataModel
+            The data model holding the data.
+        row : sequence of int
+            The row in the data model being queried.
+        column : sequence of int
+            The column in the data model being queried.
+
+        Returns
+        -------
+        text : str
+            The text to display.
+        """
+        return self.format is not None
+
+    def get_text(self, model, row, column):
+        """ Get the display text from the underlying value.
+
+        Parameters
+        ----------
+        model : AbstractDataModel
+            The data model holding the data.
+        row : sequence of int
+            The row in the data model being queried.
+        column : sequence of int
+            The column in the data model being queried.
+
+        Returns
+        -------
+        text : str
+            The text to display.
+        """
+        return self.format(model.get_value(row, column))
+
+    def has_color(self, model, row, column):
+        """ Whether or not the value has color data.
+
+        has_color : bool
+            Whether or not the value has data-associated color
+            values.
+        """
+        return (
+            self.colors is not None
+            and self.colors(model.get_value(row, column)) is not None
```

Comment:
I'd suggest this `has_color` simply return `self.colors is not None` to match `has_text`. This would be a simpler API for `colors` and this has the added benefits not having to call `model.get_value` twice when there is a valid color (once in `has_colors` and once in `get_color`).

Developers supplying the `colors` callable should make sure any value that can be obtained from the model successfully can be turned into a color. They are going to have to think about this anyway in order to return None for those scenario, so they might as well decide a default color to use.  If we document the API for this `colors` (we should), it would be simpler to say `colors` must return a color, than to say `colors` may return a color or `None`, and then have to explain what the view is going to do if the returned value is `None`.

Reply:
The reason for allowing `None` as a return value is that it meshes nicely with using `dict.get` when you have a dictionary mapping values to colors - my initial cut at this was to have the user specify a dictionary of colors (and similarly for images and text), but ended up going with a callable because it is more flexible.

I was in two minds about testing here vs. trying and failing in the getter, and I could go either way.  This is the constant tension between "look before you leap" vs. "better to ask forgiveness than permission".

Reply:
Hmmm I think I am seeing this `EnumValue` (likewise for `BoolValue` and friends), as a convenient thing that can be used as the default representations for certain types of values. On the one hand, I see there is a pull towards a simple declarative API for defining this object. On the other hand, there is a pull for flexibility to customize. And I think the two can't be met without landing us in the middle where it is neither trivial to use nor trivial to customize.  I think the most flexible approach is for anyone who want flexibility to define their own value type.

Reply:
Also, I notice that with `has_color` + `get_color` + `colors(value) -> Color or None` where `None` means no color, if I put all these together it achieves what we have in TraitsUI TabularEditor's TabularModel using a single method:
https://github.com/enthought/traitsui/blob/8f782bbdee9437d8bc6a2ba941f6d4e21cea8a3d/traitsui/qt4/tabular_model.py#L93-L94

Maybe we could just fall back to that?

Reply:
I went through that idea way back (see https://github.com/enthought/pyface/pull/543#discussion_r444024714).  It's fine for color but it breaks down for things like editor_value, and I wanted each of the channels to have a similar API.

But I am leaning towards raising if a `None` is returned by the callable, rather than checking for `None` explicitly in `has_color`.  As long as this is exceptional, rather than something which happens all the time, it should be fine.  Whatever way I go, I will make color and image follow the same convention.

Reply:
> But I am leaning towards raising DataViewGetError if a None is returned by the callable, rather than checking for None explicitly in has_color. As long as this is exceptional, rather than something which happens all the time, it should be fine. Whatever way I go, I will make color and image follow the same convention.

OK!